distributed/protocol/core.py

```diff
@@ -48,17 +48,27 @@ def dumps(msg, serializers=None, on_error="message", context=None):
         for key, (head, frames) in data.items():
             if "lengths" not in head:
                 head["lengths"] = tuple(map(nbytes, frames))
-            if "compression" not in head:
-                frames = frame_split_size(frames)
-                if frames:
-                    compression, frames = zip(*map(maybe_compress, frames))
-                else:
-                    compression = []
-                head["compression"] = compression
-            head["count"] = len(frames)
+
+            # Compress frames that are not yet compressed
+            out_compression = []
+            _out_frames = []
+            for frame, compression in zip(
+                frames, head.get("compression") or [None] * len(frames)
+            ):
+                if compression is None:  # default behavior
+                    _frames = frame_split_size(frame)
```

Comment:
I'm a little worried that this doesn't actually do what we want. Though I could be wrong.

In particular it appears when `serializer` is `"cuda"`, we [set each frame's compression to `None`]( https://github.com/dask/distributed/blob/384080422f5bd54d16ba23161a1c2d18f74ff299/distributed/protocol/cuda.py#L21 ). As a result frames will get split here giving us the behavior we wanted to avoid.

Feel free to correct me if I'm just missing something 🙂

Reply:
Maybe the answer is just we should be setting `"compression"` to `False` in that case?

Reply:
Ah indeed.  I failed to push, but yes, we want to do exactly that I think

Reply:
That seems reasonable. Thanks Matt! 😄

So I guess whenever we do decide to have compression for CUDA data, we should update for frame splitting and merging behavior accordingly. Does that sound right?

Reply:
Yeah, I think that in that situation we might register a frame size per compression type?

```python
max_frame_size = {
    "zstd": None,
    "blosc": 2**31,
    "cuda-zstd": None,
}
```

I don't know though, I think that it'll be easier once/if that happens.

Reply:
So `None` here would mean preserve existing frame size or something else?

Agree we don't need to worry about it now.

Reply:
> So None here would mean preserve existing frame size or something else?

Yeah, we only need to split frames if the compression technology requires it.